docs/codeql/codeql-language-guides/analyzing-data-flow-in-ruby.rst

```diff
@@ -0,0 +1,387 @@
+
+       /** Gets the parameter corresponding to this node, if any. */
+       Parameter asParameter() { ... }
+
+      ...
+     }
+
+You can also use the predicates ``exprNode`` and ``parameterNode``:
+
+.. code-block:: ql
+
+     /**
+      * Gets a node corresponding to expression `e`.
+      */
+     ExprNode exprNode(CfgNodes::ExprCfgNode e) { ... }
+
+     /**
+      * Gets the node corresponding to the value of parameter `p` at function entry.
+      */
+     ParameterNode parameterNode(Parameter p) { ... }
+
+Note that since ``asExpr`` and ``exprNode`` map between data-flow and control-flow nodes, you then need to call the ``getExpr`` member predicate on the control-flow node to map to the corresponding AST node,
+e.g. by writing ``node.asExpr().getExpr()``.
+Due to the control-flow graph being split, there can be multiple data-flow and control-flow nodes associated with a single expression AST node.
+
+The predicate ``localFlowStep(Node nodeFrom, Node nodeTo)`` holds if there is an immediate data flow edge from the node ``nodeFrom`` to the node ``nodeTo``.
+You can apply the predicate recursively, by using the ``+`` and ``*`` operators, or you can use the predefined recursive predicate ``localFlow``.
+
+For example, you can find flow from an expression ``source`` to an expression ``sink`` in zero or more local steps:
+
+.. code-block:: ql
+
+     DataFlow::localFlow(source, sink)
+
+Using local taint tracking
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Local taint tracking extends local data flow by including non-value-preserving flow steps.
+For example:
+
+.. code-block:: ruby
+
+     temp = x
+     y = temp + ", " + temp
+
+If ``x`` is a tainted string then ``y`` is also tainted.
+
+The local taint tracking library is in the module ``TaintTracking``.
+Like local data flow, a predicate ``localTaintStep(DataFlow::Node nodeFrom, DataFlow::Node nodeTo)`` holds if there is an immediate taint propagation edge from the node ``nodeFrom`` to the node ``nodeTo``.
+You can apply the predicate recursively, by using the ``+`` and ``*`` operators, or you can use the predefined recursive predicate ``localTaint``.
+
+For example, you can find taint propagation from an expression ``source`` to an expression ``sink`` in zero or more local steps:
+
+.. code-block:: ql
+
+     TaintTracking::localTaint(source, sink)
+
+
+Using local sources
+~~~~~~~~~~~~~~~~~~~
+
+When asking for local data flow or taint propagation between two expressions as above, you would normally constrain the expressions to be relevant to a certain investigation.
+The next section will give some concrete examples, but there is a more abstract concept that we should call out explicitly, namely that of a local source.
+
+A local source is a data-flow node with no local data flow into it.
+As such, it is a local origin of data flow, a place where a new value is created.
+This includes parameters (which only receive global data flow) and most expressions (because they are not value-preserving).
```

Comment:
Maybe?

```suggestion
This includes parameters (which only receive values from global data flow) and most expressions (because they are not value-preserving).
```